### ISSUE TITLE
Add config option for allowing extra spaces around elements and arguments in JavaScript

### ISSUE THREAD

**Issue by gziolo** [prettier/prettier#13107]
Previously proposed in #1303 and locked as resolved by bot .

## Description

This applies only to JavaScript/TypeScript code. It would be similar to the new Enforce Single Attribute Per Line option introduced with #5501 in [Prettier 2.6.0](https://prettier.io/blog/2022/03/16/2.6.0.html).

[WordPress JavaScript coding conventions for spacing](https://developer.wordpress.org/coding-standards/wordpress-coding-standards/javascript/#spacing) for historical reasons promote different formatting styles around spaces in JavaScript. The short version of the approach looks as follows:

> Use spaces liberally throughout your code. “When in doubt, space it out.”

This is the only feature that isn't supported by Prettier.

In practice, code formatting in WordPress projects look as follows:

> Always include extra spaces around elements and arguments:
>
> ```js
> array = [ a, b ];
> foo( arg );`
> foo( 'string', object );
> foo( options, object[ property ] );
> foo( node, 'property', 2 );
> prop = object[ 'default' ];
> firstArrayElement = arr[ 0 ];
>
> if ( condition ) {
>    doSomething( 'with a string' );
> } else if ( otherCondition ) {
>     otherThing( {
>         key: value,
>         otherKey: otherValue
>     } );
> } else {
>     somethingElse( true );
> }
>
> // Unlike jQuery, WordPress prefers a space after the ! negation operator.
> // This is also done to conform to our PHP standards.
> while ( ! condition ) {
>     iterating++;
> }
> 
> let i;
> for ( i = 0; i < 100; i++ ) {
>     object[ array[ i ] ] = someFn( i );
>     $( '.container' ).val( array[ i ] );
> }
>  
> try {
>     // Expressions
> } catch ( e ) {
>     // Expressions
> }
> ```

### Using Prettier's fork and npm aliases

This is where we are now for WordPress projects with [Prettier's fork that adds a `--paren-spacing` option](https://github.com/automattic/wp-prettier). We've been maintaining it for over 4 yours now.

It's possible to use all powerful integration from Prettier ecosystem with npm aliases:

https://github.com/WordPress/gutenberg/blob/d48b6bb7f2324db82ac6d3bc87b9a71e542d107f/packages/scripts/package.json#L75

```json
"dependencies": {
    "prettier": "npm:wp-prettier@2.6.2"
}
```

However npm aliases are completely unreliable in npm, and it became only worse since new behavior for peer dependencies was added in npm 7. More details in https://github.com/WordPress/gutenberg/issues/21872. I guess npm aliases are fine for a consuming project as a workaround, but using them in packages distributed on npm creates random errors.

### Changing coding conventions in WordPress projects

It was suggested in a few places in other issues in Prettier repository. It didn't seem a viable option so far despite several tries like https://github.com/WordPress/gutenberg/pull/37607 and [Proposal: Changes to JavaScript Coding Standards for Full Prettier Compatibility](https://make.wordpress.org/core/2022/01/05/proposal-changes-to-javascript-coding-standards-for-full-prettier-compatibility/).

### Plugins: allow overriding internal plugins, e.g., printer for JS or CSS

It was filed in #5919. The idea would be to create a plugin that exports a custom printer for JS/TS that replaces what Prettier offers in the core. This way, we could customize the formatted code. It isn't ideal, but that removes the maintenance work from the Prettier project for custom options like the one proposed. We've been maintaining a [Prettier's fork that adds a `--paren-spacing` option](https://github.com/automattic/wp-prettier) for the last four years. We could turn it into a plugin and publish it to NPM as `prettier-plugin-wordpress`.

**Comment by fisker** [comment]
I don't think we'll accept.

> However npm aliases are completely unreliable in npm

Will this work?
```js
// package.json
"dependencies": {
    "prettier": "link:./prettier.js"
}

// prettier.js
module.exports = require("wp-prettier");
```

**Comment by azaozz** [comment]
> I don't think we'll accept.

@fisker could you please expand a little on this? Any particular reason why parity with ESLint is bad here?

I know, lots of opinions everywhere in code formatting, but this one deals with reading ergonomics. Having more spaces does make it a tiny bit easier to read.

**Comment by kachkaev** [comment]
@azaozz see https://prettier.io/docs/en/option-philosophy.html. If you want to avoid conflicts with ESLint, please use [eslint-config-prettier](https://www.npmjs.com/package/eslint-config-prettier).

The reading ergonomics argument looks subjective. I haven’t seen any research confirming that `( x === 42 )` reads better than `(x === 42)`. I suggest closing this issue as a duplicate of https://github.com/prettier/prettier/issues/1303.

**Comment by JanJakes** [comment]
I was asked why I used 👎 for the original issue so let me share my answer also here.

As a disclaimer, I work at Automattic, but this opinion is 100% my personal one:

---

I do genuinely believe Prettier should add no such option. I think it would be harmful to both Prettier (fragmentation, violating their [philosophy](https://prettier.io/docs/en/option-philosophy.html)) and WordPress (never aligning with the rest of the ecosystem).

Therefore, I think they should not accept it. Prettier is opinionated, they always had the philosophy of never adding any configuration and I believe it’s a good philosophy. If they add options like this then it’s no longer opinionated and it’s no longer unifying the JS ecosystem. Coding styles will become fragmented again and people will discuss them forever again.

Prettier styles may not be perfect but they are very good. In the JS world, almost every larger project uses these and it would be only beneficial for WordPress to become less of a stranger to the rest of the industry.

[The issue to adopt Prettier as it is](https://make.wordpress.org/core/2022/01/05/proposal-changes-to-javascript-coding-standards-for-full-prettier-compatibility/) had huge support (19 vs. 3) and I think we should be asking why and how it ended up being thrown out of the window rather than whether Prettier should do what WordPress wants. If a popular suggestion was somehow blocked, we should be reopening a discussion with those who are blocking it. Personally, I think it's WordPress that has a lot to catch up on modern coding styles and alignment with the industry standards.

**Comment by gziolo** [comment]
> Plugins: allow overriding internal plugins, e.g., printer for JS or CSS

@fisker and @kachkaev, I mentioned that as an alternative approach. Would you reconsider https://github.com/prettier/prettier/issues/5919 that would essentially remove all the integration issues that exist today, allowing to use of a different printer for JS in the WordPress project as long as it has different handling for spacing? In fact, everything else is 100% compatible with Prettier, and when npm installs dependencies correctly, then code formatting creates amazing experience in IDEs or when using it from the terminal.

**Comment by azaozz** [comment]
> The ’reading ergonomics’ argument looks subjective...

@kachkaev yes, perhaps it is. Such small details become more important when you've been reading every day for about 40 years, and have to decrease the resolution of your monitor in order to be able to read for more than 10 minutes at a time :) 

Joking aside, I believe this is similar to typing two spaces after the period in a sentence. This is a left-over from the typewriters era, it was used to enhance the separation of the sentences. It is no longer needed as most text is displayed using a proportional font (typewriters are by default monospaced). However source code is typically displayed in a monospaced font too, so the extra spaces may improve readability in some cases/for some people.

> haven’t seen any research confirming that `( x === 42 )` reads better than `(x === 42)`

You're right, these two are quite similar. However these: `( ! something )` vs. `(!something)` are not.

**Comment by ljharb** [comment]
Since almost all JS code outside of wordpress and jquery uses the latter, it doesn't seem like the ecosystem agrees that the former is more readable.

**Comment by ljharb** [comment]
tbh, I’d argue that php and js should actively *not* be consistent, since newcomers to either might be confused about the vastly different semantics of both languages. JS should be idiomatic for JS, PHP should be idiomatic for PHP, and two distinct styles seems like it’d make learning two distinct languages much simpler.

Obv that debate doesn’t belong on this thread - but i do hope, for the sake of literally everyone that starts on JS in Wordpress (which is a lot of people) and evolves to writing non-Wordpress JS, that Wordpress eventually moves to idiomatic JS styling.

**Comment by JanJakes** [comment]
I agree that the readability part may be very subjective in this case. Also, I think it can largely be a matter of habit.

I think we shouldn't drag too much of the WP conversations here, but I feel like when already said, this should be corrected:

> we’ve tried to [convince the community](https://make.wordpress.org/core/2022/01/05/proposal-changes-to-javascript-coding-standards-for-full-prettier-compatibility/) to abandon the extra spacing for full compatibility with Prettier. But, alas, the community has rejected that proposal.

The community hasn't rejected the proposal. In fact, it was exactly the opposite — the support for this change was almost unanimous. In the proposal shared above, there were 19 people in favor of this change and 3 against it.

If one argues that many of them aren't core committers, fair enough, but note that from the core committers who shared their view, 3 were positive (Riad Benguella, Miguel Fonseca, Grzegorz Ziółkowski), 1 neutral (Felix Arntz, who is in favor as long as we also change the PHP spacing standards), and 2 negative (Aaron Jorbin, Dominik Schilling).

Therefore, I believe that statement is completely false — the community strongly supports adopting unmodified Prettier formatting, and from the core, there was also more support seen than opposition.

With such strong support, we may ask what happened that this was thrown out of the window. And I don't know — it puzzles me too. The only thing we know is [a message was shared in Slack](https://wordpress.slack.com/archives/C5UNMSU4R/p1648555187344279?thread_ts=1647357562.940759&cid=C5UNMSU4R) saying that:

> The consensus at the moment is a “No” from the core team.

Interestingly, this person is not a core committer. When asked for the context and expand on this, no answer was given (yet), although promised. That was 4 months ago. And that's what closed https://github.com/WordPress/gutenberg/pull/37607. To this day, we don't seem to know who, if anybody, made that "consensus" that goes completely against what we've seen in the public thread.

Sorry for the off-topic, but once opened, I believe the facts need to be stated.

**Comment by azaozz** [comment]
> I agree that the readability part may be very subjective

Of course it is subjective, it only affects people with poorer sight. What I am trying to draw some attention at here is that it also may be contributing (in a teeny-weeny way) to making people's sight poorer. Of course, that's unproven as there are no studies on the subject. There are some studies on what seems to be a related issue: "when using a typewriter (or a monospaced font) two spaces after the period at the end of a sentence is considered to make the text easier to read". 

That's all. Only time will tell if after 20-25 years most of the people reading this today would come to realize their sight was maybe affected. Or maybe not, who knows what reading will be like then :) 

> I think we shouldn't drag too much of the WP conversations here

Correct, this is a bad idea. 

How long have you been contributing to the WordPress open source project? Perhaps have a look at the project's philosophy: https://wordpress.org/about/philosophy/.

**Comment by ljharb** [comment]
@azaozz that was the old dogma, but in fact there are zero studies that prove that even "two spaces after a period in a monospaced font" is more readable.

**Comment by KevinBatdorf** [comment]
> How long have you been contributing to the WordPress open source project? Perhaps have a look at the project's philosophy

From the [linked document](https://wordpress.org/about/philosophy/#minority), does the "The Vocal Minority" philosophy section apply here? It feels more like a vocal minority are blocking this popular change toward aligning with the rest of the JavaScript ecosystem. Just judging by what I've read in previous conversations on the topic and chatting with other WordPress developers.

> for the sake of literally everyone that starts on JS in Wordpress (which is a lot of people) and evolves to writing non-Wordpress JS, that Wordpress eventually moves to idiomatic JS styling.

To me this seems to be the best argument in favor of WP changing coding styles. Having a unique, niche, style that few other projects use makes WordPress inherently less accessible to the broader JavaScript community (I don't believe this is an exaggeration either).

> What I am trying to draw some attention at here is that it also may be contributing (in a teeny-weeny way) to making people's sight poorer.

Maybe less compelling of an argument, I'll point out that in English we are trained to read `()` with no space, as evident with every instance in the conversation above where parenthesis were used. Reading it with the space ( case in point ) is distracting to those fluent in English.

> who knows what reading will be like then :)

Hopefully some sort of direct interface into the brain so we just process the bits directly.

**Comment by talldan** [comment]
It'd be good to not bring WordPress politics into the prettier repo. I don't think the devs here are really concerned with how decisions are made in WordPress. 😄 

The conversation is about 'should prettier adopt spaces in parentheses/braces'. The decision is likely to be made based on whether this is beneficial to consumers, how much work it is to maintain, and whether it fits into the project's goals.

I'm a WordPress contributor/maintainer, so I'm not speaking on behalf of prettier maintainers when I say this, but more from a position of respecting someone else's open source project.

**Comment by gziolo** [comment]
> The conversation is about 'should prettier adopt spaces in parentheses/braces'. The decision is likely to be made based on whether this is beneficial to consumers, how much work it is to maintain, and whether it fits into the project's goals.

I can share some more details about some of the aspects. The initial version of [`wp-prettier`](https://www.npmjs.com/package/wp-prettier) fork was published to npm 5 years ago. I believe that all necessary changes unrelated to the option were contributed by its maintainers back to this project. For 2.6.2 version, there was a single patch applied with code changes (also another one with formatting):

https://github.com/prettier/prettier/commit/5293221a2b62d5e82f6d8e896231c6c12f460711

The patch seems big at first glance, but the vast majority of changes are related to new test files:

<img width="1127" alt="Screenshot 2022-07-20 at 10 39 09" src="https://user-images.githubusercontent.com/699132/179937690-a51e6606-643b-4df1-a64c-f67192db0e60.png">

The package isn't obviously as popular as Prettier, but it still has noticeable usage:

<img width="385" alt="Screenshot 2022-07-20 at 09 28 44" src="https://user-images.githubusercontent.com/699132/179938312-0d7fcbc0-5ac9-4de2-8f52-32b3aecf2b20.png">

**Comment by homestar9** [comment]
No matter where you stand on the issue of extra spaces around parentheses `( foo )` vs `(foo)`, I think the sheer number of participants in this discussion (and many others) warrants that Prettier should add a configuration option that users can either enable or disable.  That way, people who want the extra spaces can have them, and those that do not, can continue to use Prettier as they currently do.

**Comment by nin-jin** [comment]
A few points that you should keep in mind:

- Ergonomics is not subjective. The opinions of people who do not understand ergonomics, who are not able to objectively compare the options, which makes them consider them equivalent, are subjective.

- An attempt to completely unify formatting is doomed to failure for various reasons:
   - the existence of a huge amount of code in different styles
   - the existence of projects where code ergonomics is placed above unification
   - the existence of different languages with similar syntax, where different patterns are established
  
- If a developer gets used to one style on his project, then he does not develop the skill of ignoring the style when reading code. This leads to hysterics when he have to switch to other projects instead of productive work.

I would also understand if you forced best practices in this way. But you are forcing a non-ergonomic formatting style, the only advantage of which is its popularity among projects that don’t care about ergonomics. For this reason, neither linters nor autoformatters are used in [our ecosystem](https://github.com/hyoo-ru/). Each time trying to twist them caused more problems than benefits.

Let me give you an example. Here is a simple example written by a programmer, taking into ergonomics:

```ts
class $my_super_application extends $mol_view {
  
  	@ $acme_track
	say_something_cute(
		greeting = "hello",
		name = "world",
		ending = "!",
	) {

		let message = `${ name.toUpperCase() }${ending}`
		if( message && greeting ) message = `${greeting}, ${message}`
        
		if( message ) console.log( tpl.sanitize( greeting ) )
		else console.warn( 'Empty message' )

	}
  
}
```

Now let's format it with your tool:

```ts
class $my_super_application extends $mol_view {
    @$acme_track
    say_something_cute(greeting = "hello", name = "world", ending = "!") {
        let message = `${name.toUpperCase()}${ending}`;
        if (message && greeting) message = `${greeting}, ${message}`;

        if (message) console.log(tpl.sanitize(greeting));
        else console.warn("Empty message");
    }
}
```

Absolutely everything is bad here - the [proximity principle](https://habr.com/ru/companies/logomachine/articles/324874/) is violated in every line:
- The method code does not form a single group. Instead, it breaks down the middle, and merges with the prologue and epilogue of the class.
- The separate parameters of the method were lined up in one line and also ceased to be separate groups.
- The dog stuck to the dollar, making it difficult to quickly read the [FQN](https://en.wikipedia.org/wiki/Fully_qualified_name) of function.
- It is difficult to find the boundaries of an inline expression in a template, since again the separation by spaces is lost.
- Different function calls stuck together, making it difficult to split into subexpressions.

Также рекомендую к прочтению [этот матерал](https://habr.com/ru/companies/epam_systems/articles/517398/).

**Comment by gziolo** [comment]
Since there was no progress, I’m closing this thread. I appreciate everyone’s feedback..
